Force byteable registers for indir op source

adiaaida opened the pull request:
When we have a GT_STOREIND, we need to not only make sure that the
source and dest of the storeind are marked as requiring byteable
registers, but also the indirect op's source. For example:

```
GT_STOREIND(GT_ADD(TYP_INT, mem of TYP_BYTE, reg of TYP_INT or TYP_BYTE))
```

We need to make sure that the source reg of the GT_ADD is byteable.

This change fixes #7224.

adiaaida commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs8
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8

adiaaida commented:
@sivarv PTAL
/cc: @dotnet/jit-contrib

sivarv commented on src/jit/lowerxarch.cpp:
```diff
@@ -921,11 +921,15 @@ void Lowering::TreeNodeInfoInit(GenTree* tree)
     if (varTypeIsByte(tree) || ((tree->OperGet() == GT_CAST) && varTypeIsByte(tree->CastToType())) ||
-        (tree->OperIsCompare() && varTypeIsByte(tree->gtGetOp1()) && varTypeIsByte(tree->gtGetOp2())))
+        (tree->OperIsCompare() && varTypeIsByte(tree->gtGetOp1()) && varTypeIsByte(tree->gtGetOp2())) ||
+        ((tree->OperGet() == GT_ADD) && varTypeIsByte(tree->gtGetOp1()) && (tree->gtGetOp1()->OperGet() == GT_IND)))
```
We need a more holistic fix than this.  Think of GT_SUB/GT_MUL node with similar IR.

I think we need to revisit how we can correctly set srcCandidates on operands confined to bytable regs where required.  

We need to examine to see if we can come up with a set of conditions that would characterize all the cases where bytable regs are required.  If it is not possible to come up with such a characterization, another option to examine is whether each GT_* node should set srcCandidates individually.

adiaaida commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs8
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8

adiaaida commented:
@sivarv I've updated this PR to move the condition to TreeNodeInfoInitLogicalOp. I've also updated TreeNodeInfoInitCmp to handle the issue opened by @pgavlin earlier today (#7478). It is more restrictive than we originally discussed, because I wanted to also try to catch the case that @BruceForstall mentioned in #7483. I wonder if we should be more restrictive now to eliminate the asserts/silent bad codegen, and open an issue to look into being less restrictive.

mikedn commented on src/jit/lowerxarch.cpp:
```diff
@@ -3511,6 +3523,20 @@ void Lowering::TreeNodeInfoInitCmp(GenTreePtr tree)
+            regMask = op1->gtLsraInfo.getSrcCandidates(m_lsra);
+            assert(regMask != RBM_NONE);
+            op1->gtLsraInfo.setSrcCandidates(m_lsra, regMask & ~RBM_NON_BYTE_REGS);
+        }
```
Should this also set the destination candidates? The xor from #7483 (and pretty much any other binary op) needs both operands to be byte registers.

sivarv replied:
A compare node that produces a result in a register requires its destination candidates be a byteable reg because it uses "setcc  bytereg" instruction.

mikedn commented:
> I wonder if we should be more restrictive now to eliminate the asserts/silent bad codegen, and open an issue to look into being less restrictive.

Or just disable the optimization on x86. It saves a cheap `movzx` but risks introducing more spilling due the very limited number of byte registers.

adiaaida commented on src/jit/lowerxarch.cpp:
```diff
@@ -2299,6 +2299,18 @@ void Lowering::TreeNodeInfoInitLogicalOp(GenTree* tree)
+            regMask = op2->gtLsraInfo.getSrcCandidates(l);
+            assert(regMask != RBM_NONE);
+            op2->gtLsraInfo.setSrcCandidates(l, regMask & ~RBM_NON_BYTE_REGS);
+        }
```
Do we only care if it's a mem-op? Or would we want to constrain op1 if op2 is TYP_BYTE no matter what?

sivarv replied:
I think the right condition to check for logical operators (add,sub, and, or, xor) is

if (varTypeIsByte(op1) || varTypeIsByte(op2))
{
     // Set source candidates of op1 and op2 to be byteable registers
     // Set destination candidates of tree to be byteable registers
}

Note that the above logic would be more restrictive than required in the following case
GT_ADD(op1= byte ptr mem-op, op2=TYP_INT or TYP_BYTE in a reg)

We just need to constrain dst candiates of GT_ADD to byteable regs and not op2.
During codegen, if op2's Reg == targetReg, then op2 is already in a byteable reg.
If op2's reg != targetReg,  codegen will do a reg-to-reg move using genActualType(targetType).

CarolEidt replied:
I think that we need to be crisper about the invariants here. I, at least, am a bit confused about what cases we will encounter here. I think (I could be mistaken, and haven't investigated) that TYP_BYTE or TYP_UBYTE lclVars (and many subexpressions involving byte types) are effectively promoted to TYP_INT, with casts added by morph as needed to ensure proper semantics. Therefore, when we see these operations, they generally will be targeting something that is probably going to be stored to memory. @sivarv do I have that right? Otherwise, this would seem to be overly conservative because it is not checking whether we are in a situation that actually requires a byteable register.

---
In reply to: [81855970](https://github.com/dotnet/coreclr/pull/7466#discussion_r81855970) [](ancestors = 81855970)

pgavlin replied:
> 2) Locals that are normalized on store, should have their type normalized to int or long by importer.

I think that the condition here is actually "lclVars that are not normalized on load": AFAICT the importer never calls `lvNormalizeOnStore`.

Incidentally, does anybody know why args are considered normalize-on-load but not normalize-on-store? I can understand the former--I assume that we do load normalization to guard against the caller not doing store-normalization--but I'm a bit lost as to the former (unless the arg is also aliasable).

CarolEidt replied:
I believe (and I don't feel that I grok this 100%) it is the case that these are duals - i.e. you never have both. The idea is that you count on either having a full integer-width entity that you can load from and store into (normalize on store), OR you have to load from and store into a byte and zero- or sign-extend as needed on load (because it's a field of a local struct or an arg). In the former case you need to make sure that you only store a byte's worth, even if you have somehow accumulated garbage in the upper bits. In the latter case, it doesn't matter because you're only storing a byte. I hope somebody will clue me in if I don't have this right!

adiaaida commented:
@sivarv PTAL

sivarv commented on src/jit/lowerxarch.cpp:
```diff
@@ -4087,6 +4087,18 @@ bool Lowering::SetStoreIndOpCountsIfRMWMemOp(GenTreePtr storeInd)
+    if (varTypeIsByte(storeInd))
+    {
+        // If storeInd is of TYP_BYTE, set indirOpSources to byteable registers.
+        bool containedNode = (indirOpSource->gtLsraInfo.srcCount == 0) && (indirOpSource->gtLsraInfo.dstCount == 0);
```
Here we just need to check dstCount of indirOpSource is zero.  if its dstCount == 0 implies it is marked as contained.

adiaaida replied:
Fixed.

sivarv commented:
Looks good.

CarolEidt commented:
LGTM

CarolEidt commented on src/jit/lowerxarch.cpp:
```diff
@@ -4087,6 +4087,18 @@ bool Lowering::SetStoreIndOpCountsIfRMWMemOp(GenTreePtr storeInd)
+            assert(regMask != RBM_NONE);
+            indirOpSource->gtLsraInfo.setSrcCandidates(m_lsra, regMask & ~RBM_NON_BYTE_REGS);
+        }
+    }
```
Good catch, Siva!

adiaaida replied:
Good catch! Fixed.

adiaaida commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs8
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8

sivarv commented:
Since the root cause of this issue turned out to be different than what was initially thought, please update the description of this issue.

adiaaida commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs8
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8
